consensus-types/hdiff/state_diff.go

```diff
@@ -0,0 +1,2101 @@
+		Slot:          primitives.Slot(binary.LittleEndian.Uint64((*data)[:8])),
+		ProposerIndex: primitives.ValidatorIndex(binary.LittleEndian.Uint64((*data)[8:16])),
+		ParentRoot:    slices.Clone((*data)[16 : 16+fieldparams.RootLength]),
+		StateRoot:     slices.Clone((*data)[16+fieldparams.RootLength : 16+2*fieldparams.RootLength]),
+		BodyRoot:      slices.Clone((*data)[16+2*fieldparams.RootLength : 16+3*fieldparams.RootLength]),
+	}
+	*data = (*data)[blockHeaderLength:]
+	return nil
+}
+
+func (ret *stateDiff) readBlockRoots(data *[]byte) error {
+	if len(*data) < blockRootsLength {
+		return errors.Wrap(errDataSmall, "blockRoots")
+	}
+	for i := range fieldparams.BlockRootsLength {
+		copy(ret.blockRoots[i][:], (*data)[i*fieldparams.RootLength:(i+1)*fieldparams.RootLength])
+	}
+	*data = (*data)[blockRootsLength:]
+	return nil
+}
+
+func (ret *stateDiff) readStateRoots(data *[]byte) error {
+	if len(*data) < stateRootsLength {
+		return errors.Wrap(errDataSmall, "stateRoots")
+	}
+	for i := range fieldparams.StateRootsLength {
+		copy(ret.stateRoots[i][:], (*data)[i*fieldparams.RootLength:(i+1)*fieldparams.RootLength])
+	}
+	*data = (*data)[stateRootsLength:]
+	return nil
+}
+
+func (ret *stateDiff) readHistoricalRoots(data *[]byte) error {
+	if len(*data) < 8 {
+		return errors.Wrap(errDataSmall, "historicalRoots")
+	}
+	historicalRootsLength := int(binary.LittleEndian.Uint64((*data)[:8])) // lint:ignore uintcast
+	(*data) = (*data)[8:]
+	if len(*data) < historicalRootsLength*fieldparams.RootLength {
+		return errors.Wrap(errDataSmall, "historicalRoots")
+	}
+	ret.historicalRoots = make([][fieldparams.RootLength]byte, historicalRootsLength)
+	for i := range historicalRootsLength {
+		copy(ret.historicalRoots[i][:], (*data)[i*fieldparams.RootLength:(i+1)*fieldparams.RootLength])
+	}
+	*data = (*data)[historicalRootsLength*fieldparams.RootLength:]
+	return nil
+}
+
+func (ret *stateDiff) readEth1Data(data *[]byte) error {
+	if len(*data) < 1 {
+		return errors.Wrap(errDataSmall, "eth1Data")
+	}
+	if (*data)[0] == nilMarker {
+		*data = (*data)[1:]
+		return nil
+	}
+	*data = (*data)[1:]
+	if len(*data) < eth1DataLength {
+		return errors.Wrap(errDataSmall, "eth1Data")
+	}
+	ret.eth1Data = &ethpb.Eth1Data{
+		DepositRoot:  slices.Clone((*data)[1 : 1+fieldparams.RootLength]),
+		DepositCount: binary.LittleEndian.Uint64((*data)[1+fieldparams.RootLength : 1+fieldparams.RootLength+8]),
+		BlockHash:    slices.Clone((*data)[1+fieldparams.RootLength+8 : 1+2*fieldparams.RootLength+8]),
+	}
+	*data = (*data)[1+eth1DataLength:]
+	return nil
+}
+
+func (ret *stateDiff) readEth1DataVotes(data *[]byte) error {
+	// Read eth1DataVotes.
+	if len(*data) < 9 {
+		return errors.Wrap(errDataSmall, "eth1DataVotes")
+	}
+	ret.eth1VotesAppend = ((*data)[0] == nilMarker)
+	eth1DataVotesLength := int(binary.LittleEndian.Uint64((*data)[1 : 1+8])) // lint:ignore uintcast
+	if len(*data) < 1+8+eth1DataVotesLength*eth1DataLength {
```

Comment:
maybe we should do a max() check either 9 or 1+8+eth1DataVotesLength*eth1DataLength whichever is longer?

Reply:
The length cannot be known until after checking for the first 8 bytes. Since the nilmarker anyway needs to be present, that's why I check for 9 instead of just 9